File: Rust/rust_aoc/src/year_2015/d16_aunt_sue.rs
use std::collections::HashMap;
use regex::Regex;

pub fn part_one(lines: &Vec<String>) -> String {
    solve(lines, is_exact_match).to_string()
}

pub fn part_two(lines: &Vec<String>) -> String {
    solve(lines, is_range_match).to_string()
}

type Sue = HashMap<String, u32>;

fn get_target_sue() -> Sue {
    let mut result: Sue = HashMap::new();
    result.insert(String::from("children"), 3);
    result.insert(String::from("cats"), 7);
    result.insert(String::from("samoyeds"), 2);
    result.insert(String::from("pomeranians"), 3);
    result.insert(String::from("akitas"), 0);
    result.insert(String::from("vizslas"), 0);
    result.insert(String::from("goldfish"), 5);
    result.insert(String::from("trees"), 3);
    result.insert(String::from("cars"), 2);
    result.insert(String::from("perfumes"), 1);

    result
}

fn parse_sue(line: &str) -> Sue {
    //Sue 1: goldfish: 6, trees: 9, akitas: 0
    let re = Regex::new(r#"(\w+): (\d+)"#).unwrap();

    let mut result = Sue::new();

    re.captures_iter(line).map(|c| {
        let (_, [name, value]) = c.extract();
        (name, value)
    }).for_each(|(name, value)| {
        result.insert(String::from(name), value.parse::<u32>().unwrap());
    });

    result
}

fn is_exact_match(sue: &Sue, target: &Sue) -> bool {
    !target.iter().any(|t| sue.contains_key(t.0) && sue.get(t.0).unwrap() != t.1)
}

fn is_range_match(sue: &Sue, target: &Sue) -> bool {
    for (key, value) in target.iter() {
        if !sue.contains_key(key) {
            continue;
        }

        if  key == "cats" || key == "trees" {
            if sue.get(key).unwrap() <= value {
                return false;
            } 
        }
        else if key == "pomeranians" || key == "goldfish" {
            if sue.get(key).unwrap() >= value {
                 return false;
            }
        }
        else {
            if sue.get(key).unwrap() != value {
                return false;
            }
        }
    }

    true
}

fn solve(lines: &Vec<String>, matcher: fn(&Sue, &Sue)->bool) -> usize {
    let target = get_target_sue();
    for i in 0..lines.len() {
        let sue = parse_sue(&lines[i]);
        if matcher(&sue, &target) {
            return i + 1;
        }
    }
    0
}

File: Cpp/AdventOfCode/src/2015/d16_AuntSue.cpp
#pragma once

#include "2015/d16_AuntSue.h"

SOLUTION(2015, 16) {

    using Sue = Constexpr::SmallMap<std::string, u32>;

    constexpr Sue GetTargetSue() {
        Sue result;
        result["children"] = 3;
        result["cats"] = 7;
        result["samoyeds"] = 2;
        result["pomeranians"] = 3;
        result["akitas"] = 0;
        result["vizslas"] = 0;
        result["goldfish"] = 5;
        result["trees"] = 3;
        result["cars"] = 2;
        result["perfumes"] = 1;

        return result;
    }
    constexpr Sue ParseSue(std::string_view line) {
        //Sue 1: goldfish: 6, trees: 9, akitas: 0

        auto propStart = line.find(':');
        auto rhs = line.substr(propStart + 2);
        auto s1 = Constexpr::Split(rhs, ", ");

        Sue result;
        for (const auto& pair : s1) {
            auto s2 = Constexpr::Split(pair, ": ");
            auto prop = std::string(s2[0]);
            auto valStr = std::string(s2[1]);
            u32 val;
            Constexpr::ParseNumber(valStr, val);
            result[prop] = val;
        }

        return result;
    }

    constexpr bool IsExactMatch(const Sue & sue) {
        const auto TargetSue = GetTargetSue();
        for (const auto& [key, value] : TargetSue) {
            if (sue.contains(key) && sue.at(key) != value) return false;
        }

        return true;
    }

    constexpr bool IsRangeMatch(const Sue & sue) {
        const auto TargetSue = GetTargetSue();
        for (const auto& [key, value] : TargetSue) {
            if (!sue.contains(key)) continue;

            if (key == "cats" || key == "trees") {
                if (sue.at(key) <= value) return false;
            }
            else if (key == "pomeranians" || key == "goldfish") {
                if (sue.at(key) >= value) return false;
            }
            else {
                if (sue.at(key) != value) return false;
            }
        }
        return true;
    }

    constexpr size_t Solve(const auto& lines, auto match) {
        auto sues = ParseLines(lines, ParseSue);
        for (size_t i = 0; i < sues.size(); i++) {
            if (match(sues[i])) return i + 1;
        }
        return 0;
    }
    PART_ONE() {
        return Constexpr::ToString(Solve(lines, IsExactMatch));
    }
    PART_TWO() {
        return Constexpr::ToString(Solve(lines, IsRangeMatch));
    }

    TESTS() {
        return true;
    }
}